File: src/storage/lazy_store.py
from __future__ import annotations

from collections.abc import Callable
from datetime import datetime

from src.domain.models import JSONValue, VariableRecord, VariableSetItem
from src.storage.store import VariableStore


class LazyVariableStore:
    def __init__(self, factory: Callable[[], VariableStore]) -> None:
        self._factory = factory
        self._store: VariableStore | None = None

    def get(self, key: str) -> VariableRecord | None:
        return self._get_store().get(key)

    def set(
        self,
        key: str,
        value: JSONValue,
        expires_at: datetime | None = None,
    ) -> VariableRecord:
        return self._get_store().set(key, value, expires_at=expires_at)

    def delete(self, key: str) -> VariableRecord | None:
        return self._get_store().delete(key)

    def list(self, prefix: str | None = None) -> list[VariableRecord]:
        return self._get_store().list(prefix=prefix)

    def bulk_set(self, items: list[VariableSetItem]) -> list[VariableRecord]:
        return self._get_store().bulk_set(items)

    def bulk_delete(self, keys: list[str]) -> list[str]:
        return self._get_store().bulk_delete(keys)

    def export_snapshot(self) -> dict[str, JSONValue]:
        return self._get_store().export_snapshot()

    def import_snapshot(self, data: dict[str, JSONValue]) -> dict[str, JSONValue]:
        return self._get_store().import_snapshot(data)

    def _get_store(self) -> VariableStore:
        if self._store is None:
            self._store = self._factory()
        return self._store

File: src/storage/lazy-store.ts
import { JSONValue, VariableRecord, VariableSetItem } from "../domain/models";
import { VariableStore } from "./store";

export class LazyVariableStore implements VariableStore {
  private readonly factory: () => VariableStore;
  private store: VariableStore | null = null;

  constructor(factory: () => VariableStore) {
    this.factory = factory;
  }

  async get(key: string): Promise<VariableRecord | null> {
    return this.getStore().get(key);
  }

  async set(
    key: string,
    value: JSONValue,
    expiresAt?: string | null,
  ): Promise<VariableRecord> {
    return this.getStore().set(key, value, expiresAt);
  }

  async delete(key: string): Promise<VariableRecord | null> {
    return this.getStore().delete(key);
  }

  async list(prefix?: string | null): Promise<VariableRecord[]> {
    return this.getStore().list(prefix);
  }

  async bulkSet(items: VariableSetItem[]): Promise<VariableRecord[]> {
    return this.getStore().bulkSet(items);
  }

  async bulkDelete(keys: string[]): Promise<string[]> {
    return this.getStore().bulkDelete(keys);
  }

  async exportSnapshot(): Promise<Record<string, JSONValue>> {
    return this.getStore().exportSnapshot();
  }

  async importSnapshot(data: Record<string, JSONValue>): Promise<Record<string, JSONValue>> {
    return this.getStore().importSnapshot(data);
  }

  private getStore(): VariableStore {
    if (this.store == null) {
      this.store = this.factory();
    }
    return this.store;
  }
}
